google/cloud/bigtable/_mutate_rows.py

```diff
@@ -0,0 +1,239 @@
+            # add this exception to list for each active mutation
+            for idx in error_dict.keys():
+                error_dict[idx].append(exc)
+            # remove non-idempotent mutations from mutations_dict, so they are not retried
+            for idx, mut in list(mutations_dict.items()):
+                if not mut.is_idempotent():
+                    mutations_dict.pop(idx)
+
+    retry = retries.AsyncRetry(
+        predicate=predicate,
+        on_error=on_error_fn,
+        timeout=operation_timeout,
+        initial=0.01,
+        multiplier=2,
+        maximum=60,
+    )
+    # use generator to lower per-attempt timeout as we approach operation_timeout deadline
+    attempt_timeout_gen = _attempt_timeout_generator(
+        per_request_timeout, operation_timeout
+    )
+    # wrap attempt in retry logic
+    retry_wrapped = retry(_mutate_rows_retryable_attempt)
+    # convert RetryErrors from retry wrapper into DeadlineExceeded errors
+    deadline_wrapped = _convert_retry_deadline(retry_wrapped, operation_timeout)
+    try:
+        # trigger mutate_rows
+        await deadline_wrapped(
+            gapic_client,
+            table,
+            attempt_timeout_gen,
+            mutations_dict,
+            error_dict,
+            predicate,
+            on_terminal_state,
+        )
+    except Exception as exc:
+        # exceptions raised by retryable are added to the list of exceptions for all unfinalized mutations
+        for error_list in error_dict.values():
+            error_list.append(exc)
+    finally:
+        # raise exception detailing incomplete mutations
+        all_errors = []
+        for idx, exc_list in error_dict.items():
+            if len(exc_list) == 0:
+                raise core_exceptions.ClientError(
+                    f"Mutation {idx} failed with no associated errors"
+                )
+            elif len(exc_list) == 1:
+                cause_exc = exc_list[0]
+            else:
+                cause_exc = bt_exceptions.RetryExceptionGroup(exc_list)
+            entry = mutation_entries[idx]
+            all_errors.append(
+                bt_exceptions.FailedMutationEntryError(idx, entry, cause_exc)
+            )
+            # call on_terminal_state for each unreported failed mutation
+            if on_terminal_state:
+                await on_terminal_state(idx, entry, exc_list)
+        if all_errors:
+            raise bt_exceptions.MutationsExceptionGroup(
+                all_errors, len(mutation_entries)
+            )
+
+
+async def _mutate_rows_retryable_attempt(
+    gapic_client: "BigtableAsyncClient",
+    table: "Table",
+    timeout_generator: Iterator[float],
+    active_dict: dict[int, "RowMutationEntry"],
+    error_dict: dict[int, list[Exception]],
+    predicate: Callable[[Exception], bool],
+    on_terminal_state: Callable[
+        [int, "RowMutationEntry", list[Exception] | None],
+        Coroutine[None, None, None],
+    ]
+    | None = None,
```

Comment:
It seems like you are abusing bound args as state variables. I think the intent of this function are much better expressed as a class with instance variables to represent state.

```
class _RetryAttempt:
   def __init__(self, gapic, mutations, rpc_timeout, operation_timeout):
      self.gapic = gapic
      self.mutations = mutations
      self.remaining_indices = range(0, len(mutations))
      self.errors = {}
      self.rpc_timeout = rpc_timeout
      self.operation_timeout = operation_timeout

   def run(self):
       request = { ... }
       for i in self.remaining_indices:
          request["mutations"].append(self.mutations[i])

      new_remaining_indices = []
      try:
          results = gapic.mutate_rows(request)
          for r in results:
              orig_index = self.remaining_indices[r.index]
              entry_error = from_grpc_status(r.error)
            
              if entry_error is None:
                  pass
              else:
                  self.errors[orig_index].append(entry_error)
                  if is_idempotent(self.mutations[orig_index]) and  is_retriable(entry_error):
                      new_remaining_indices.append(orig_index)
        catch e:
            # handle full rpc error

        self.remaining_indices = new_remaining_indices
        if self.remaining_indices:
            raise IncompleteError
```

Then the operation function can instantiate it, bind its run method to the instance and decorate it with AsyncRetry